### PR TITLE
✨ Add `is7BitASCII(s, allowed)` to LibString

### PR DESCRIPTION
**Pull request by atarpara**
## Description
closing #962 

## Checklist

Ensure you completed **all of the steps** below before submitting your pull request:

- [x] Ran `forge fmt`?
- [x] Ran `forge snapshot`?
- [x] Ran `forge test`?

_Pull requests with an incomplete checklist will be thrown out._

### CODE REVIEW

File: src/utils/LibString.sol [hunk 1]
```diff
@@ -406,6 +406,36 @@ library LibString {
         }
     }
 
+    /// @dev Returns if this string is a alphanumeric character including space.
```
Comment (PaulRBerg) [inline src/utils/LibString.sol:409]: For clarity, I would rewrite the NatSpec to:

```
/// @dev Returns true if the string contains only alphanumeric characters and spaces.
```

File: src/utils/LibString.sol [hunk 2]
```diff
@@ -406,6 +406,36 @@ library LibString {
...
+                        break
+                    }
+                    o := add(o, result)
+                    if iszero(lt(o, end)) { break }
+                }
+            }
+        }
+    }
+
+    /// @dev Returns if this character is a alphanumeric including space.
```
Comment (PaulRBerg) [inline src/utils/LibString.sol:431]: For clarity, I would rewrite the NatSpec to:

```
/// @dev Returns true if the character is an alphanumeric character or a space.
```

**Review by PaulRBerg — APPROVED** [review]
very nice, thank you

**Comment by 0xCLARITY** [comment]
Left a more detailed comment on the issue: https://github.com/Vectorized/solady/issues/962#issuecomment-2176574758, but I think it would be _surprising_ for spaces to be considered "alphanumeric".

**Comment by Vectorized** [comment]
Refactored to make it more general.

Will merge in 24 hours if no hard objections.

**Review by Vectorized — APPROVED** [review]

**Comment by atarpara** [comment]
@Vectorized In the current implementation of the `is7BitASCII(string memory s, uint128 allowed)` function, The behavior for empty strings is to return true in all cases. This is inconsistent with the behavior observed in similar functions across various programming languages. 

For instance:

Python [isalnum](https://github.com/python/cpython/blob/main/Objects/clinic/unicodeobject.c.h#L603-L604)
Rust's [is_ascii_alphanumeric](https://doc.rust-lang.org/beta/std/primitive.char.html#method.is_ascii_alphanumeric)
Govalidator's [alphanumeric](https://github.com/asaskevich/govalidator/blob/master/patterns.go#L16)

In these implementations, the function typically returns false for an empty string. To align with these standards and ensure consistent behavior, I propose the following adjustments:

1. The function will return false for empty strings, aligning with the default behavior in other languages.
2. If allowing empty strings is necessary for specific use cases, a separate function can be created. Alternatively, developers can handle this explicitly by checking the length of the string using `bytes(str).length == 0`.

Everyone is open to give your valuable suggestions.

**Comment by Vectorized** [comment]
Let's deviate away from the popular languages.

I find it easier to reason if an empty string evaluates to true.

**Pull request merged**